---
title: "Conchola_2022_bioinformatics"
author: "Zhiwei Xiao"
date: "6/13/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# general fetal data processing
```{r}
library(Seurat)

#navigate to appropriate folder for each 10x run.
f59ddistal.data <- Read10X(data.dir = "./")
f59dairway.data <- Read10X(data.dir = "./")
f132ddistal.data <- Read10X(data.dir = "./")
f80ddistal.data <- Read10X(data.dir = "./")
f80dairway.data <- Read10X(data.dir = "./")
f145dtrach.data <- Read10X(data.dir = "./")
f103ddistal.data <- Read10X(data.dir = "./")
f103dairway.data <- Read10X(data.dir = "./")
f103dtrach.data <- Read10X(data.dir = "./")
f125dairway.data <- Read10X(data.dir = "./")
f125ddistal.data <- Read10X(data.dir = "./")

f103dairway <- CreateSeuratObject(counts = f103dairway.data, project = "f103dairway", min.cells = 3, min.features = 200)
f103ddistal <- CreateSeuratObject(counts = f103ddistal.data, project = "f103ddistal", min.cells = 3, min.features = 200)
f125dairway <- CreateSeuratObject(counts = f125dairway.data, project = "f125dairway", min.cells = 3, min.features = 200)
f132ddistal <- CreateSeuratObject(counts = f132ddistal.data, project = "f132ddistal", min.cells = 3, min.features = 200)
f125ddistal <- CreateSeuratObject(counts = f125ddistal.data, project = "f125ddistal", min.cells = 3, min.features = 200)
f59dairway <- CreateSeuratObject(counts = f59dairway.data, project = "f59dairway", min.cells = 3, min.features = 200)
f59ddistal <- CreateSeuratObject(counts = f59ddistal.data, project = "f59ddistal", min.cells = 3, min.features = 200)
f80dairway  <- CreateSeuratObject(counts = f80dairway.data, project = "f80dairway", min.cells = 3, min.features = 200)
f80ddistal <- CreateSeuratObject(counts = f80ddistal.data, project = "f80ddistal", min.cells = 3, min.features = 200)
f145dtrach <- CreateSeuratObject(counts = f145dtrach.data, project = "f145dtrach", min.cells = 3, min.features = 200)
f103dtrach <- CreateSeuratObject(counts = f103dtrach.data, project = "f103dtrach", min.cells = 3, min.features = 200)


f103dairway <- RenameCells(object = f103dairway, add.cell.id = "f103dairway")
f103ddistal <- RenameCells(object = f103ddistal, add.cell.id = "f103ddistal")
f125dairway <- RenameCells(object = f125dairway, add.cell.id = "f125dairway")
f132ddistal <- RenameCells(object = f132ddistal, add.cell.id = "f132ddistal")
f125ddistal <- RenameCells(object = f125ddistal, add.cell.id = "f125ddistal")
f59dairway <- RenameCells(object = f59dairway, add.cell.id = "f59dairway")
f59ddistal <- RenameCells(object = f59ddistal, add.cell.id = "f59ddistal")
f80dairway  <- RenameCells(object = f80dairway, add.cell.id = "f80dairway")
f80ddistal <- RenameCells(object = f80ddistal, add.cell.id = "f80ddistal")
f145dtrach <- RenameCells(object = f145dtrach, add.cell.id = "f145dtrach")
f103dtrach <- RenameCells(object = f103dtrach, add.cell.id = "f103dtrach")

f103dairway[["percent.mt"]] <- PercentageFeatureSet(f103dairway, pattern = "^MT-") 
f103ddistal[["percent.mt"]] <- PercentageFeatureSet(f103ddistal, pattern = "^MT-") 
f125dairway[["percent.mt"]] <- PercentageFeatureSet(f125dairway, pattern = "^MT-") 
f132ddistal[["percent.mt"]] <- PercentageFeatureSet(f132ddistal, pattern = "^MT-") 
f125ddistal[["percent.mt"]] <- PercentageFeatureSet(f125ddistal, pattern = "^MT-")
f59dairway[["percent.mt"]] <- PercentageFeatureSet(f59dairway, pattern = "^MT-") 
f59ddistal[["percent.mt"]] <- PercentageFeatureSet(f59ddistal, pattern = "^MT-") 
f80dairway[["percent.mt"]]  <- PercentageFeatureSet(f80dairway, pattern = "^MT-") 
f80ddistal[["percent.mt"]] <- PercentageFeatureSet(f80ddistal, pattern = "^MT-")
f145dtrach[["percent.mt"]] <- PercentageFeatureSet(f145dtrach, pattern = "^MT-")
f103dtrach[["percent.mt"]] <- PercentageFeatureSet(f103dtrach, pattern = "^MT-")


fetal.combined = merge(f103dairway, y = c(f103ddistal, f125dairway, f132ddistal, f125ddistal, f59dairway, f59ddistal, f80dairway, f80ddistal, f103dtrach, f145dtrach), project = "fetallung")
ribo.genes <- grep(pattern = "^RP[SL][[:digit:]]", x = rownames(x = fetal.combined), value = TRUE);
percent.ribo <- Matrix::colSums(x = GetAssayData(object = fetal.combined, slot = 'counts')[ribo.genes, ]) / Matrix::colSums(x = GetAssayData(object = fetal.combined, slot = 'counts'));
fetal.combined[['percent.ribo']] <- percent.ribo;

setwd("~/scRNAseq/Output/2021-08-22 Bud-Tip to BTA/All Data")


pdf(file.path("./", paste0("QC percent MT", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "percent.mt")
dev.off()
pdf(file.path("./", paste0("QC nFeature_RNA", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "nFeature_RNA")
dev.off()
pdf(file.path("./", paste0("QC percent ribo", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "percent.ribo")
dev.off()
pdf(file.path("./", paste0("QC nCount_RNA", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "nCount_RNA", y.max = 50000)
dev.off()

VlnPlot(fetal.combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
fetal.combined <- subset(fetal.combined, subset = nFeature_RNA > 500 & nFeature_RNA < 12000 & percent.mt < 10)

fetal.combined <- SplitObject(fetal.combined, split.by = "orig.ident")
for (i in seq_along(fetal.combined)) {
  fetal.combined[[i]] <- NormalizeData(fetal.combined[[i]]) %>% FindVariableFeatures()
}
features <- SelectIntegrationFeatures(fetal.combined)
for (i in seq_along(along.with = fetal.combined)) {
  fetal.combined[[i]] <- ScaleData(fetal.combined[[i]], features = features) %>% RunPCA(features = features)
}
anchors <- FindIntegrationAnchors(fetal.combined, anchor.features = features, reduction = "rpca", dims = 1:30)
fetal.combined.integrated <- IntegrateData(anchors, dims = 1:30)
DefaultAssay(fetal.combined.integrated) <- "integrated"
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
fetal.combined.integrated <- CellCycleScoring(fetal.combined.integrated, s.features = s.genes, g2m.features = g2m.genes)
fetal.combined.integrated <- ScaleData(fetal.combined.integrated, vars.to.regress = c("S.Score", "G2M.Score"))
fetal.combined.integrated <- RunPCA(fetal.combined.integrated)
ElbowPlot(fetal.combined.integrated, ndims = 50)
fetal.combined.integrated <- RunUMAP(fetal.combined.integrated, dims = 1:30, reduction.name = "umap", return.model = TRUE)
fetal.combined.integrated <- FindNeighbors(fetal.combined.integrated, dims = 1:30)
fetal.combined.integrated <- FindClusters(fetal.combined.integrated, resolution = 0.5)
DimPlot(fetal.combined.integrated, group.by = c("orig.ident", "ident"), label = TRUE, pt.size = 0.5)


compartment <- c("VIM", "PTPRC", "CDH1", "EPCAM", "PECAM1", "PDGFRA", "CDH5")
epithelial.markers <- c("SFTPC","SFTPA1", "ABCA3", "HOPX", "AGER", "SOX9", "SOX2", "FGF20", "TP63", "FOXJ1", "CHGA", 'SCGB3A2')


pdf(file.path("./", paste0("Louvain", ".pdf")), w=11, h=8.5)
DimPlot(fetal.combined.integrated, reduction = "umap", label = TRUE, pt.size = 2)
dev.off()
pdf(file.path("./", paste0("UMAPbySample", ".pdf")), w=11, h=8.5)
DimPlot(fetal.combined.integrated, reduction = "umap", group.by = "orig.ident", label = TRUE, pt.size = 2)
dev.off()





DefaultAssay(fetal.combined.integrated) <- "RNA"
fetal.combined.integrated <- NormalizeData(fetal.combined.integrated)


pdf(file.path("./", paste0("Compartment", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.combined.integrated, features = compartment, pt.size = 0.2, order = TRUE)
dev.off()

pdf(file.path("./", paste0("Proliferating", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.combined.integrated, features = c("MKI67", "TOP2A"), pt.size = 0.2, order = TRUE)
dev.off()

pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.combined.integrated, features = epithelial.markers, pt.size = 0.2, order = TRUE)
dev.off()


#extract epithelium based on above 3 feature plots and enrichment lists
setwd("~/scRNAseq/Output/2021-08-22 Bud-Tip to BTA/Epithelial")

fetal.epithelial <- subset(fetal.combined.integrated, idents = c(13, 23, 10, 12, 5, 19, 22))
fetal.epithelial.cellids <- WhichCells(fetal.epithelial)

DefaultAssay(fetal.epithelial) <- "integrated"
fetal.epithelial <- RunPCA(fetal.epithelial, verbose = FALSE)
ElbowPlot(fetal.epithelial, ndims = 50)
fetal.epithelial <- RunUMAP(fetal.epithelial, reduction = "pca", dims = 1:12, return.model = TRUE)
fetal.epithelial <- FindNeighbors(fetal.epithelial, dims = 1:12)
fetal.epithelial <- FindClusters(fetal.epithelial, resolution = 0.2)
DimPlot(fetal.epithelial, group.by = c("orig.ident", "ident"), label = TRUE, pt.size = 0.5)

#Figure 1A
pdf(file.path("./", paste0("Louvain", ".pdf")), w=11, h=8.5)
DimPlot(fetal.epithelial, reduction = "umap", label = TRUE, pt.size = 2)
dev.off()

#Figure S1D
pdf(file.path("./", paste0("UMAPbySample", ".pdf")), w=11, h=8.5)
DimPlot(fetal.epithelial, reduction = "umap", group.by = "orig.ident", label = TRUE, pt.size = 2)
dev.off()





DefaultAssay(fetal.epithelial) <- "RNA"
fetal.epithelial <- NormalizeData(fetal.epithelial)

pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.epithelial, features = epithelial.markers, pt.size = 2, order = TRUE)
dev.off()

#Figure 1B
pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.epithelial, features = c("SCGB3A2", "SFTPB", "CFTR"), pt.size = 2, order = TRUE)
dev.off()
#Figure 3B
pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.epithelial, features = c("C6", "MUC16"), pt.size = 0.2, order = TRUE)
dev.off()
#Data for Figure S1E
write.csv(table(Idents(fetal.epithelial)), "CellsPerCluster.csv") #number of cells in each cluster
write.csv(table(Idents(fetal.epithelial), fetal.epithelial$orig.ident), "ClusterbyRun.csv")

zero.markers <- FindMarkers(fetal.epithelial, ident.1 = 0, min.pct = 0.25)
one.markers <- FindMarkers(fetal.epithelial, ident.1 = 1, min.pct = 0.25)
two.markers <- FindMarkers(fetal.epithelial, ident.1 = 2, min.pct = 0.25)
three.markers <- FindMarkers(fetal.epithelial, ident.1 = 3, min.pct = 0.25)
four.markers <- FindMarkers(fetal.epithelial, ident.1 = 4, min.pct = 0.25)
five.markers <- FindMarkers(fetal.epithelial, ident.1 = 5, min.pct = 0.25)
six.markers <- FindMarkers(fetal.epithelial, ident.1 = 6, min.pct = 0.25)
seven.markers <- FindMarkers(fetal.epithelial, ident.1 = 7, min.pct = 0.25)
eight.markers <- FindMarkers(fetal.epithelial, ident.1 = 8, min.pct = 0.25)
nine.markers <- FindMarkers(fetal.epithelial, ident.1 = 9, min.pct = 0.25)

write.csv(zero.markers, "Cluster0.csv")
write.csv(one.markers, "Cluster1.csv")
write.csv(two.markers, "Cluster2.csv")
write.csv(three.markers, "Cluster3.csv")
write.csv(four.markers, "Cluster4.csv")
write.csv(five.markers, "Cluster5.csv")
write.csv(six.markers, "Cluster6.csv")
write.csv(seven.markers, "Cluster7.csv")
write.csv(eight.markers, "Cluster8.csv")
write.csv(nine.markers, "Cluster9.csv")

```

# general cell tag data processing
```{r}
library(CellTagR)
##set working directory to bam file
bam.test.obj <- CellTagObject(object.name = "bam.cell.tag.obj", fastq.bam.directory = "./ubam/")
bam.test.obj <- CellTagExtraction(bam.test.obj, celltag.version = "v1")
bam.test.obj.unprocessed <- bam.test.obj
head(bam.test.obj@bam.parse.rslt[["v1"]])
setwd("~/scRNAseq/Data/CellTag/2021ACSHT4??CellTagV 1Sort/barcodes")
bam.test.obj <- CellTagMatrixCount(celltag.obj = bam.test.obj, barcodes.file = "./barcodes.tsv")
dim(bam.test.obj@raw.count)
setwd("~/scRNAseq/Data/CellTag/2021ACSHT4??CellTagV 1Sort/collapsing")
bam.test.obj.nostarcode <- bam.test.obj
bam.test.obj <- CellTagDataForCollapsing(celltag.obj = bam.test.obj, output.file = "collapsing.txt")
##move collapsing.txt to the desktop folder (delete old files) and run the following in terminal from the starcode folder:
## cd..
## cd starcode
## cd starcode
## ./starcode -s --print-clusters ~/Desktop/collapsing.txt > ~/Desktop/collapsing_result.txt
##move collapsing_result file back to collapsing directoryhead(bam.test.obj@collapsed.count)
bam.test.obj <- CellTagDataPostCollapsing(celltag.obj = bam.test.obj, collapsed.rslt.file = "./collapsing_result.txt")
bam.test.obj <- SingleCellDataBinatization(bam.test.obj, 2)
bam.test.obj.nowhitelist <- bam.test.obj

#set working directory to whitelist file
bam.test.obj <- SingleCellDataWhitelist(bam.test.obj, whitels.cell.tag.file = "./v1_whitelist.csv")

bam.test.obj.prefilter <- bam.test.obj
bam.test.obj <- MetricBasedFiltering(bam.test.obj, 2, comparison = "greater")
bam.test.obj <- MetricBasedFiltering(bam.test.obj, 20, comparison = "less")
bam.test.obj <- JaccardAnalysis(bam.test.obj)
bam.test.obj <- CloneCalling(celltag.obj = bam.test.obj, correlation.cutoff=0.7)
bam.test.obj@clone.size.info[["v1"]]
bam.test.obj@clone.composition[["v1"]]
bam.test.obj.clonescalled <- bam.test.obj 

library(Seurat)
#setwd(source for data)
celltag.data <- Read10X(data.dir = "./")
celltag <- CreateSeuratObject(counts = celltag.data, project = "RUN1", min.cells = 3, min.features = 200)
cloneid <- bam.test.obj@clone.composition[["v1"]][["clone.id"]]
clonebarcode <- bam.test.obj@clone.composition[["v1"]][["cell.barcode"]]
cloneid2 <- data.frame(cloneid, row.names = clonebarcode)
celltag <- AddMetaData(object = celltag, metadata = cloneid2, col.name = "cloneid")


celltag[["percent.mt"]] <- PercentageFeatureSet(celltag, pattern = "^MT-")
VlnPlot(celltag, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
dev.off()
celltag <- subset(celltag, subset = nFeature_RNA > 500 & nFeature_RNA < 10000 & percent.mt < 10)
celltag <- NormalizeData(celltag)
celltag <- FindVariableFeatures(celltag, selection.method = "vst", nfeatures = 500)
celltag <- ScaleData(celltag)
celltag <- RunPCA(celltag, features = VariableFeatures(celltag))
ElbowPlot(celltag, ndims = 50)
celltag <- RunUMAP(celltag, dims = 1:10, reduction.name = "umap", return.model = TRUE)
celltag <- FindNeighbors(celltag, dims = 1:10)
celltag <- FindClusters(celltag, resolution = 0.3)

#Figure 1C
pdf(file.path("./", paste0("Cell Tag Louvain", ".pdf")), w=11, h=8.5)
DimPlot(celltag, label = TRUE, pt.size = 2)
dev.off()

pdf(file.path("./", paste0("Cell Tag Louvain No Label", ".pdf")), w=11, h=8.5)
DimPlot(celltag, label = FALSE, pt.size = 2)
dev.off()
DefaultAssay(celltag) <- "RNA"
celltag <- NormalizeData(celltag)

pdf(file.path("./", paste0("FASrootedclonesV3", ".pdf")), w=11, h=8.5)
DimPlot(celltag, cells.highlight = fasrooted.cellids, pt.size = 2, sizes.highlight = 3)
dev.off()
pdf(file.path("./", paste0("BasalrootedclonesV3", ".pdf")), w=11, h=8.5)
DimPlot(celltag, cells.highlight = basalrooted.cellids, pt.size = 2, sizes.highlight = 3)
dev.off()

write.csv(table(Idents(celltag)), "CellsPerCluster.csv") #number of cells in each cluster
write.csv(table(Idents(celltag), celltag$orig.ident), "ClusterbyRun.csv") #number of cells in each cluster by run
highfrequencyclones <- bam.test.obj@clone.size.info[["v1"]] 
highfreqcloneid <-subset(highfrequencyclones, Frequency >= 10, select = Clone.ID) #identify clones with frequency >= 10
celltag@meta.data["cloneid"][is.na(celltag@meta.data["cloneid"])] <- 0 #set all NA clones to 0 for subsetting
for (i in highfreqcloneid[["Clone.ID"]]) {
  cells.use <- colnames(celltag)[which(celltag[[]][["cloneid"]] == c(i))]
  pdf(file.path("./", paste0("clone", i, ".pdf")), w=11, h=8.5)
  print(DimPlot(celltag, reduction = "umap", cells.highlight = cells.use, sizes.highlight = 3, pt.size = 2, order = TRUE))
  dev.off()
} #export a pdf of the UMAP for every clone equal or greater than 10 in frequency

#export a heatmap of clonesbycluster
clonebycluster <- table(celltag$cloneid, Idents(celltag)) #number of cells in each cluster by cloneid
clonebycluster.rownames <- rownames(clonebycluster)
clonebyclustermtx <- as.data.frame.matrix(clonebycluster)
clonebyclustermtx <- mutate(clonebyclustermtx, Frequency = rowSums(clonebyclustermtx))
rownames(clonebyclustermtx) <- clonebycluster.rownames
clonebyclustermtxhighfreq <- subset(clonebyclustermtx, Frequency >= 10)
clonebyclustermtx <- as.matrix(clonebyclustermtxhighfreq)
clonebyclustermtx <- clonebyclustermtx[,-9]
clonebyclustermtx <- clonebyclustermtx[-1,]
write.csv(clonebyclustermtx, "CloneByClusterMtx.csv")
pdf(file.path("./", paste0("ClonebyClusterHeatmap", ".pdf")), w=11, h=8.5)
heatmap(clonebyclustermtx)
dev.off()


##output enrichment lists
zero.markers <- FindMarkers(celltag, ident.1 = 0, min.pct = 0.25)
one.markers <- FindMarkers(celltag, ident.1 = 1, min.pct = 0.25)
two.markers <- FindMarkers(celltag, ident.1 = 2, min.pct = 0.25)
three.markers <- FindMarkers(celltag, ident.1 = 3, min.pct = 0.25)
four.markers <- FindMarkers(celltag, ident.1 = 4, min.pct = 0.25)
five.markers <- FindMarkers(celltag, ident.1 = 5, min.pct = 0.25)
six.markers <- FindMarkers(celltag, ident.1 = 6, min.pct = 0.25)
seven.markers <- FindMarkers(celltag, ident.1 = 7, min.pct = 0.25)

write.csv(zero.markers, "Cluster0.csv")
write.csv(one.markers, "Cluster1.csv")
write.csv(two.markers, "Cluster2.csv")
write.csv(three.markers, "Cluster3.csv")
write.csv(four.markers, "Cluster4.csv")
write.csv(five.markers, "Cluster5.csv")
write.csv(six.markers, "Cluster6.csv")
write.csv(seven.markers, "Cluster7.csv")


##create cellid lists of each rooted clone type
basalrooted.subset <-subset(celltag, subset = cloneid == "3" | cloneid == "7" | cloneid == "10" | cloneid == "30" | cloneid == "47" | cloneid == "60" | cloneid == "64" | cloneid == "73" | cloneid == "80" | cloneid == "93" | cloneid == "95" | cloneid == "101" | cloneid == "108") 
basalrooted.cellids <- WhichCells(basalrooted.subset)
fasrooted.subset <- subset(celltag, subset = cloneid == "2" | cloneid == "4" | cloneid == "5" | cloneid == "11" | cloneid == "12" | cloneid == "14" | cloneid == "20" | cloneid == "21" | cloneid == "24" | cloneid == "26" | cloneid == "28" | cloneid == "29" | cloneid == "38" | cloneid == "42" | cloneid == "43" | cloneid == "53" | cloneid == "56" | cloneid == "71" | cloneid == "79" | cloneid == "92" | cloneid == "96" | cloneid == "98")
fasrooted.cellids <- WhichCells(fasrooted.subset)
#Figure 2D
pdf(file.path("./", paste0("Basal Dominant clones" ,".pdf")), w=11, h=8.5)
DimPlot(basalrooted.subset, pt.size = 2)
dev.off()
pdf(file.path("./", paste0("FAS Dominant clones", ".pdf")), w=11, h=8.5)
DimPlot(fasrooted.subset, pt.size = 2)
dev.off()

fasrooted.subset <- AddMetaData(fasrooted.subset, "fas", col.name = "root")
basalrooted.subset <- AddMetaData(basalrooted.subset, "basal", col.name = "root")
celltag <- RenameIdents(object = celltag, '0' = "Basal", '1' = "FAS", '2' = "Multiciliated", '3' = "SCGB1A1/3A1/SPDEF", '4' = "Unknown (ASCL1+)", '5' = "Cycling", '6' = "Deuterosome", '7' = "Neuroendocrine")
```




