---
title: "Conchola_2022_bioinformatics"
author: "Zhiwei Xiao"
date: "6/13/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# General fetal data processing
```{r}
library(Seurat)

#navigate to appropriate folder for each 10x run.
f59ddistal.data <- Read10X(data.dir = "./")
f59dairway.data <- Read10X(data.dir = "./")
f132ddistal.data <- Read10X(data.dir = "./")
f80ddistal.data <- Read10X(data.dir = "./")
f80dairway.data <- Read10X(data.dir = "./")
f145dtrach.data <- Read10X(data.dir = "./")
f103ddistal.data <- Read10X(data.dir = "./")
f103dairway.data <- Read10X(data.dir = "./")
f103dtrach.data <- Read10X(data.dir = "./")
f125dairway.data <- Read10X(data.dir = "./")
f125ddistal.data <- Read10X(data.dir = "./")

f103dairway <- CreateSeuratObject(counts = f103dairway.data, project = "f103dairway", min.cells = 3, min.features = 200)
f103ddistal <- CreateSeuratObject(counts = f103ddistal.data, project = "f103ddistal", min.cells = 3, min.features = 200)
f125dairway <- CreateSeuratObject(counts = f125dairway.data, project = "f125dairway", min.cells = 3, min.features = 200)
f132ddistal <- CreateSeuratObject(counts = f132ddistal.data, project = "f132ddistal", min.cells = 3, min.features = 200)
f125ddistal <- CreateSeuratObject(counts = f125ddistal.data, project = "f125ddistal", min.cells = 3, min.features = 200)
f59dairway <- CreateSeuratObject(counts = f59dairway.data, project = "f59dairway", min.cells = 3, min.features = 200)
f59ddistal <- CreateSeuratObject(counts = f59ddistal.data, project = "f59ddistal", min.cells = 3, min.features = 200)
f80dairway  <- CreateSeuratObject(counts = f80dairway.data, project = "f80dairway", min.cells = 3, min.features = 200)
f80ddistal <- CreateSeuratObject(counts = f80ddistal.data, project = "f80ddistal", min.cells = 3, min.features = 200)
f145dtrach <- CreateSeuratObject(counts = f145dtrach.data, project = "f145dtrach", min.cells = 3, min.features = 200)
f103dtrach <- CreateSeuratObject(counts = f103dtrach.data, project = "f103dtrach", min.cells = 3, min.features = 200)


f103dairway <- RenameCells(object = f103dairway, add.cell.id = "f103dairway")
f103ddistal <- RenameCells(object = f103ddistal, add.cell.id = "f103ddistal")
f125dairway <- RenameCells(object = f125dairway, add.cell.id = "f125dairway")
f132ddistal <- RenameCells(object = f132ddistal, add.cell.id = "f132ddistal")
f125ddistal <- RenameCells(object = f125ddistal, add.cell.id = "f125ddistal")
f59dairway <- RenameCells(object = f59dairway, add.cell.id = "f59dairway")
f59ddistal <- RenameCells(object = f59ddistal, add.cell.id = "f59ddistal")
f80dairway  <- RenameCells(object = f80dairway, add.cell.id = "f80dairway")
f80ddistal <- RenameCells(object = f80ddistal, add.cell.id = "f80ddistal")
f145dtrach <- RenameCells(object = f145dtrach, add.cell.id = "f145dtrach")
f103dtrach <- RenameCells(object = f103dtrach, add.cell.id = "f103dtrach")

f103dairway[["percent.mt"]] <- PercentageFeatureSet(f103dairway, pattern = "^MT-") 
f103ddistal[["percent.mt"]] <- PercentageFeatureSet(f103ddistal, pattern = "^MT-") 
f125dairway[["percent.mt"]] <- PercentageFeatureSet(f125dairway, pattern = "^MT-") 
f132ddistal[["percent.mt"]] <- PercentageFeatureSet(f132ddistal, pattern = "^MT-") 
f125ddistal[["percent.mt"]] <- PercentageFeatureSet(f125ddistal, pattern = "^MT-")
f59dairway[["percent.mt"]] <- PercentageFeatureSet(f59dairway, pattern = "^MT-") 
f59ddistal[["percent.mt"]] <- PercentageFeatureSet(f59ddistal, pattern = "^MT-") 
f80dairway[["percent.mt"]]  <- PercentageFeatureSet(f80dairway, pattern = "^MT-") 
f80ddistal[["percent.mt"]] <- PercentageFeatureSet(f80ddistal, pattern = "^MT-")
f145dtrach[["percent.mt"]] <- PercentageFeatureSet(f145dtrach, pattern = "^MT-")
f103dtrach[["percent.mt"]] <- PercentageFeatureSet(f103dtrach, pattern = "^MT-")


fetal.combined = merge(f103dairway, y = c(f103ddistal, f125dairway, f132ddistal, f125ddistal, f59dairway, f59ddistal, f80dairway, f80ddistal, f103dtrach, f145dtrach), project = "fetallung")
ribo.genes <- grep(pattern = "^RP[SL][[:digit:]]", x = rownames(x = fetal.combined), value = TRUE);
percent.ribo <- Matrix::colSums(x = GetAssayData(object = fetal.combined, slot = 'counts')[ribo.genes, ]) / Matrix::colSums(x = GetAssayData(object = fetal.combined, slot = 'counts'));
fetal.combined[['percent.ribo']] <- percent.ribo;

setwd("~/scRNAseq/Output/2021-08-22 Bud-Tip to BTA/All Data")


pdf(file.path("./", paste0("QC percent MT", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "percent.mt")
dev.off()
pdf(file.path("./", paste0("QC nFeature_RNA", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "nFeature_RNA")
dev.off()
pdf(file.path("./", paste0("QC percent ribo", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "percent.ribo")
dev.off()
pdf(file.path("./", paste0("QC nCount_RNA", ".pdf")), w=11, h=8.5)
VlnPlot(fetal.combined, features = "nCount_RNA", y.max = 50000)
dev.off()

VlnPlot(fetal.combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
fetal.combined <- subset(fetal.combined, subset = nFeature_RNA > 500 & nFeature_RNA < 12000 & percent.mt < 10)

fetal.combined <- SplitObject(fetal.combined, split.by = "orig.ident")
for (i in seq_along(fetal.combined)) {
  fetal.combined[[i]] <- NormalizeData(fetal.combined[[i]]) %>% FindVariableFeatures()
}
features <- SelectIntegrationFeatures(fetal.combined)
for (i in seq_along(along.with = fetal.combined)) {
  fetal.combined[[i]] <- ScaleData(fetal.combined[[i]], features = features) %>% RunPCA(features = features)
}
anchors <- FindIntegrationAnchors(fetal.combined, anchor.features = features, reduction = "rpca", dims = 1:30)
fetal.combined.integrated <- IntegrateData(anchors, dims = 1:30)
DefaultAssay(fetal.combined.integrated) <- "integrated"
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
fetal.combined.integrated <- CellCycleScoring(fetal.combined.integrated, s.features = s.genes, g2m.features = g2m.genes)
fetal.combined.integrated <- ScaleData(fetal.combined.integrated, vars.to.regress = c("S.Score", "G2M.Score"))
fetal.combined.integrated <- RunPCA(fetal.combined.integrated)
ElbowPlot(fetal.combined.integrated, ndims = 50)
fetal.combined.integrated <- RunUMAP(fetal.combined.integrated, dims = 1:30, reduction.name = "umap", return.model = TRUE)
fetal.combined.integrated <- FindNeighbors(fetal.combined.integrated, dims = 1:30)
fetal.combined.integrated <- FindClusters(fetal.combined.integrated, resolution = 0.5)
DimPlot(fetal.combined.integrated, group.by = c("orig.ident", "ident"), label = TRUE, pt.size = 0.5)


compartment <- c("VIM", "PTPRC", "CDH1", "EPCAM", "PECAM1", "PDGFRA", "CDH5")
epithelial.markers <- c("SFTPC","SFTPA1", "ABCA3", "HOPX", "AGER", "SOX9", "SOX2", "FGF20", "TP63", "FOXJ1", "CHGA", 'SCGB3A2')


pdf(file.path("./", paste0("Louvain", ".pdf")), w=11, h=8.5)
DimPlot(fetal.combined.integrated, reduction = "umap", label = TRUE, pt.size = 2)
dev.off()
pdf(file.path("./", paste0("UMAPbySample", ".pdf")), w=11, h=8.5)
DimPlot(fetal.combined.integrated, reduction = "umap", group.by = "orig.ident", label = TRUE, pt.size = 2)
dev.off()





DefaultAssay(fetal.combined.integrated) <- "RNA"
fetal.combined.integrated <- NormalizeData(fetal.combined.integrated)


pdf(file.path("./", paste0("Compartment", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.combined.integrated, features = compartment, pt.size = 0.2, order = TRUE)
dev.off()

pdf(file.path("./", paste0("Proliferating", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.combined.integrated, features = c("MKI67", "TOP2A"), pt.size = 0.2, order = TRUE)
dev.off()

pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.combined.integrated, features = epithelial.markers, pt.size = 0.2, order = TRUE)
dev.off()


#extract epithelium based on above 3 feature plots and enrichment lists
setwd("~/scRNAseq/Output/2021-08-22 Bud-Tip to BTA/Epithelial")

fetal.epithelial <- subset(fetal.combined.integrated, idents = c(13, 23, 10, 12, 5, 19, 22))
fetal.epithelial.cellids <- WhichCells(fetal.epithelial)

DefaultAssay(fetal.epithelial) <- "integrated"
fetal.epithelial <- RunPCA(fetal.epithelial, verbose = FALSE)
ElbowPlot(fetal.epithelial, ndims = 50)
fetal.epithelial <- RunUMAP(fetal.epithelial, reduction = "pca", dims = 1:12, return.model = TRUE)
fetal.epithelial <- FindNeighbors(fetal.epithelial, dims = 1:12)
fetal.epithelial <- FindClusters(fetal.epithelial, resolution = 0.2)
DimPlot(fetal.epithelial, group.by = c("orig.ident", "ident"), label = TRUE, pt.size = 0.5)

#Figure 1A
pdf(file.path("./", paste0("Louvain", ".pdf")), w=11, h=8.5)
DimPlot(fetal.epithelial, reduction = "umap", label = TRUE, pt.size = 2)
dev.off()

#Figure S1D
pdf(file.path("./", paste0("UMAPbySample", ".pdf")), w=11, h=8.5)
DimPlot(fetal.epithelial, reduction = "umap", group.by = "orig.ident", label = TRUE, pt.size = 2)
dev.off()





DefaultAssay(fetal.epithelial) <- "RNA"
fetal.epithelial <- NormalizeData(fetal.epithelial)

pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.epithelial, features = epithelial.markers, pt.size = 2, order = TRUE)
dev.off()

#Figure 1B
pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.epithelial, features = c("SCGB3A2", "SFTPB", "CFTR"), pt.size = 2, order = TRUE)
dev.off()
#Figure 3B
pdf(file.path("./", paste0("Epithelial Markers", ".pdf")), w=11, h=8.5)
FeaturePlot(fetal.epithelial, features = c("C6", "MUC16"), pt.size = 0.2, order = TRUE)
dev.off()
#Data for Figure S1E
write.csv(table(Idents(fetal.epithelial)), "CellsPerCluster.csv") #number of cells in each cluster
write.csv(table(Idents(fetal.epithelial), fetal.epithelial$orig.ident), "ClusterbyRun.csv")

zero.markers <- FindMarkers(fetal.epithelial, ident.1 = 0, min.pct = 0.25)
one.markers <- FindMarkers(fetal.epithelial, ident.1 = 1, min.pct = 0.25)
two.markers <- FindMarkers(fetal.epithelial, ident.1 = 2, min.pct = 0.25)
three.markers <- FindMarkers(fetal.epithelial, ident.1 = 3, min.pct = 0.25)
four.markers <- FindMarkers(fetal.epithelial, ident.1 = 4, min.pct = 0.25)
five.markers <- FindMarkers(fetal.epithelial, ident.1 = 5, min.pct = 0.25)
six.markers <- FindMarkers(fetal.epithelial, ident.1 = 6, min.pct = 0.25)
seven.markers <- FindMarkers(fetal.epithelial, ident.1 = 7, min.pct = 0.25)
eight.markers <- FindMarkers(fetal.epithelial, ident.1 = 8, min.pct = 0.25)
nine.markers <- FindMarkers(fetal.epithelial, ident.1 = 9, min.pct = 0.25)

write.csv(zero.markers, "Cluster0.csv")
write.csv(one.markers, "Cluster1.csv")
write.csv(two.markers, "Cluster2.csv")
write.csv(three.markers, "Cluster3.csv")
write.csv(four.markers, "Cluster4.csv")
write.csv(five.markers, "Cluster5.csv")
write.csv(six.markers, "Cluster6.csv")
write.csv(seven.markers, "Cluster7.csv")
write.csv(eight.markers, "Cluster8.csv")
write.csv(nine.markers, "Cluster9.csv")

```

# General celltag data processing
```{r}
library(CellTagR)
##set working directory to bam file
bam.test.obj <- CellTagObject(object.name = "bam.cell.tag.obj", fastq.bam.directory = "./ubam/")
bam.test.obj <- CellTagExtraction(bam.test.obj, celltag.version = "v1")
bam.test.obj.unprocessed <- bam.test.obj
head(bam.test.obj@bam.parse.rslt[["v1"]])
setwd("~/scRNAseq/Data/CellTag/2021ACSHT4??CellTagV 1Sort/barcodes")
bam.test.obj <- CellTagMatrixCount(celltag.obj = bam.test.obj, barcodes.file = "./barcodes.tsv")
dim(bam.test.obj@raw.count)
setwd("~/scRNAseq/Data/CellTag/2021ACSHT4??CellTagV 1Sort/collapsing")
bam.test.obj.nostarcode <- bam.test.obj
bam.test.obj <- CellTagDataForCollapsing(celltag.obj = bam.test.obj, output.file = "collapsing.txt")
##move collapsing.txt to the desktop folder (delete old files) and run the following in terminal from the starcode folder:
## cd..
## cd starcode
## cd starcode
## ./starcode -s --print-clusters ~/Desktop/collapsing.txt > ~/Desktop/collapsing_result.txt
##move collapsing_result file back to collapsing directoryhead(bam.test.obj@collapsed.count)
bam.test.obj <- CellTagDataPostCollapsing(celltag.obj = bam.test.obj, collapsed.rslt.file = "./collapsing_result.txt")
bam.test.obj <- SingleCellDataBinatization(bam.test.obj, 2)
bam.test.obj.nowhitelist <- bam.test.obj

#set working directory to whitelist file
bam.test.obj <- SingleCellDataWhitelist(bam.test.obj, whitels.cell.tag.file = "./v1_whitelist.csv")

bam.test.obj.prefilter <- bam.test.obj
bam.test.obj <- MetricBasedFiltering(bam.test.obj, 2, comparison = "greater")
bam.test.obj <- MetricBasedFiltering(bam.test.obj, 20, comparison = "less")
bam.test.obj <- JaccardAnalysis(bam.test.obj)
bam.test.obj <- CloneCalling(celltag.obj = bam.test.obj, correlation.cutoff=0.7)
bam.test.obj@clone.size.info[["v1"]]
bam.test.obj@clone.composition[["v1"]]
bam.test.obj.clonescalled <- bam.test.obj 

library(Seurat)
#setwd(source for data)
celltag.data <- Read10X(data.dir = "./")
celltag <- CreateSeuratObject(counts = celltag.data, project = "RUN1", min.cells = 3, min.features = 200)
cloneid <- bam.test.obj@clone.composition[["v1"]][["clone.id"]]
clonebarcode <- bam.test.obj@clone.composition[["v1"]][["cell.barcode"]]
cloneid2 <- data.frame(cloneid, row.names = clonebarcode)
celltag <- AddMetaData(object = celltag, metadata = cloneid2, col.name = "cloneid")


celltag[["percent.mt"]] <- PercentageFeatureSet(celltag, pattern = "^MT-")
VlnPlot(celltag, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
dev.off()
celltag <- subset(celltag, subset = nFeature_RNA > 500 & nFeature_RNA < 10000 & percent.mt < 10)
celltag <- NormalizeData(celltag)
celltag <- FindVariableFeatures(celltag, selection.method = "vst", nfeatures = 500)
celltag <- ScaleData(celltag)
celltag <- RunPCA(celltag, features = VariableFeatures(celltag))
ElbowPlot(celltag, ndims = 50)
celltag <- RunUMAP(celltag, dims = 1:10, reduction.name = "umap", return.model = TRUE)
celltag <- FindNeighbors(celltag, dims = 1:10)
celltag <- FindClusters(celltag, resolution = 0.3)

#Figure 1C
pdf(file.path("./", paste0("Cell Tag Louvain", ".pdf")), w=11, h=8.5)
DimPlot(celltag, label = TRUE, pt.size = 2)
dev.off()

pdf(file.path("./", paste0("Cell Tag Louvain No Label", ".pdf")), w=11, h=8.5)
DimPlot(celltag, label = FALSE, pt.size = 2)
dev.off()
DefaultAssay(celltag) <- "RNA"
celltag <- NormalizeData(celltag)

pdf(file.path("./", paste0("FASrootedclonesV3", ".pdf")), w=11, h=8.5)
DimPlot(celltag, cells.highlight = fasrooted.cellids, pt.size = 2, sizes.highlight = 3)
dev.off()
pdf(file.path("./", paste0("BasalrootedclonesV3", ".pdf")), w=11, h=8.5)
DimPlot(celltag, cells.highlight = basalrooted.cellids, pt.size = 2, sizes.highlight = 3)
dev.off()

write.csv(table(Idents(celltag)), "CellsPerCluster.csv") #number of cells in each cluster
write.csv(table(Idents(celltag), celltag$orig.ident), "ClusterbyRun.csv") #number of cells in each cluster by run
highfrequencyclones <- bam.test.obj@clone.size.info[["v1"]] 
highfreqcloneid <-subset(highfrequencyclones, Frequency >= 10, select = Clone.ID) #identify clones with frequency >= 10
celltag@meta.data["cloneid"][is.na(celltag@meta.data["cloneid"])] <- 0 #set all NA clones to 0 for subsetting
for (i in highfreqcloneid[["Clone.ID"]]) {
  cells.use <- colnames(celltag)[which(celltag[[]][["cloneid"]] == c(i))]
  pdf(file.path("./", paste0("clone", i, ".pdf")), w=11, h=8.5)
  print(DimPlot(celltag, reduction = "umap", cells.highlight = cells.use, sizes.highlight = 3, pt.size = 2, order = TRUE))
  dev.off()
} #export a pdf of the UMAP for every clone equal or greater than 10 in frequency

#export a heatmap of clonesbycluster
clonebycluster <- table(celltag$cloneid, Idents(celltag)) #number of cells in each cluster by cloneid
clonebycluster.rownames <- rownames(clonebycluster)
clonebyclustermtx <- as.data.frame.matrix(clonebycluster)
clonebyclustermtx <- mutate(clonebyclustermtx, Frequency = rowSums(clonebyclustermtx))
rownames(clonebyclustermtx) <- clonebycluster.rownames
clonebyclustermtxhighfreq <- subset(clonebyclustermtx, Frequency >= 10)
clonebyclustermtx <- as.matrix(clonebyclustermtxhighfreq)
clonebyclustermtx <- clonebyclustermtx[,-9]
clonebyclustermtx <- clonebyclustermtx[-1,]
write.csv(clonebyclustermtx, "CloneByClusterMtx.csv")
pdf(file.path("./", paste0("ClonebyClusterHeatmap", ".pdf")), w=11, h=8.5)
heatmap(clonebyclustermtx)
dev.off()


##output enrichment lists
zero.markers <- FindMarkers(celltag, ident.1 = 0, min.pct = 0.25)
one.markers <- FindMarkers(celltag, ident.1 = 1, min.pct = 0.25)
two.markers <- FindMarkers(celltag, ident.1 = 2, min.pct = 0.25)
three.markers <- FindMarkers(celltag, ident.1 = 3, min.pct = 0.25)
four.markers <- FindMarkers(celltag, ident.1 = 4, min.pct = 0.25)
five.markers <- FindMarkers(celltag, ident.1 = 5, min.pct = 0.25)
six.markers <- FindMarkers(celltag, ident.1 = 6, min.pct = 0.25)
seven.markers <- FindMarkers(celltag, ident.1 = 7, min.pct = 0.25)

write.csv(zero.markers, "Cluster0.csv")
write.csv(one.markers, "Cluster1.csv")
write.csv(two.markers, "Cluster2.csv")
write.csv(three.markers, "Cluster3.csv")
write.csv(four.markers, "Cluster4.csv")
write.csv(five.markers, "Cluster5.csv")
write.csv(six.markers, "Cluster6.csv")
write.csv(seven.markers, "Cluster7.csv")


##create cellid lists of each rooted clone type
basalrooted.subset <-subset(celltag, subset = cloneid == "3" | cloneid == "7" | cloneid == "10" | cloneid == "30" | cloneid == "47" | cloneid == "60" | cloneid == "64" | cloneid == "73" | cloneid == "80" | cloneid == "93" | cloneid == "95" | cloneid == "101" | cloneid == "108") 
basalrooted.cellids <- WhichCells(basalrooted.subset)
fasrooted.subset <- subset(celltag, subset = cloneid == "2" | cloneid == "4" | cloneid == "5" | cloneid == "11" | cloneid == "12" | cloneid == "14" | cloneid == "20" | cloneid == "21" | cloneid == "24" | cloneid == "26" | cloneid == "28" | cloneid == "29" | cloneid == "38" | cloneid == "42" | cloneid == "43" | cloneid == "53" | cloneid == "56" | cloneid == "71" | cloneid == "79" | cloneid == "92" | cloneid == "96" | cloneid == "98")
fasrooted.cellids <- WhichCells(fasrooted.subset)
#Figure 2D
pdf(file.path("./", paste0("Basal Dominant clones" ,".pdf")), w=11, h=8.5)
DimPlot(basalrooted.subset, pt.size = 2)
dev.off()
pdf(file.path("./", paste0("FAS Dominant clones", ".pdf")), w=11, h=8.5)
DimPlot(fasrooted.subset, pt.size = 2)
dev.off()

fasrooted.subset <- AddMetaData(fasrooted.subset, "fas", col.name = "root")
basalrooted.subset <- AddMetaData(basalrooted.subset, "basal", col.name = "root")
celltag <- RenameIdents(object = celltag, '0' = "Basal", '1' = "FAS", '2' = "Multiciliated", '3' = "SCGB1A1/3A1/SPDEF", '4' = "Unknown (ASCL1+)", '5' = "Cycling", '6' = "Deuterosome", '7' = "Neuroendocrine")
```

# FAS cell extraction
```{r}
library(Seurat)
library(tidyverse)
library(patchwork)
library(data.table)
library(hdf5r)
library(readxl)
library(leiden)
library(sctransform)
library(slingshot)
library(SingleCellExperiment)
library(RColorBrewer)
library(rgl)
library(grDevices)
library(destiny)
library(plotly)
library(SeuratWrappers)

fetal.epithelial.cluster0 <- subset(x = fetal.epithelial, idents = c(0))

DefaultAssay(fetal.epithelial.cluster0) <- "RNA"
set.seed(888)
fetal.epithelial.cluster0 <- NormalizeData(fetal.epithelial.cluster0, 
                                            normalization.method = "LogNormalize", scale.factor =10000)

fetal.epithelial.cluster0 <-  FindVariableFeatures(fetal.epithelial.cluster0)
fetal.epithelial.cluster0 <- RunFastMNN(object.list = SplitObject(fetal.epithelial.cluster0, split.by = "orig.ident"))
fetal.epithelial.cluster0 <- RunUMAP(fetal.epithelial.cluster0, dims = 1:30,reduction = "mnn")
fetal.epithelial.cluster0 <- FindNeighbors(fetal.epithelial.cluster0, dims = 1:30,reduction = "mnn")
fetal.epithelial.cluster0 <- FindClusters(fetal.epithelial.cluster0, resolution = 0.3, algorithm = 4)

```


# Celltag cluster 2 extraction (clone only)
```{r}
DefaultAssay(celltag) <- "RNA"
celltag <- NormalizeData(celltag,normalization.method = "LogNormalize", scale.factor =10000)

celltag$clone_or_not <- case_when(is.na(celltag$cloneid) ~ 0,
                               !is.na(celltag$cloneid) ~ 1)
celltag2_clone_only <- subset(x = celltag, subset = seurat_clusters == "2" & clone_or_not == 1)

set.seed(888)
celltag2_clone_only <- SCTransform(celltag2_clone_only, vars.to.regress = "percent.mt", verbose = FALSE)
celltag2_clone_only <- RunPCA(celltag2_clone_only, verbose = FALSE)
celltag2_clone_only <- RunUMAP(celltag2_clone_only, dims = 1:30)
celltag2_clone_only <- FindNeighbors(celltag2_clone_only, dims = 1:30)
celltag2_clone_only <- FindClusters(celltag2_clone_only, resolution = 0.3, algorithm = 4)

DefaultAssay(celltag2_clone_only) <- "RNA"
celltag2_clone_only <- NormalizeData(celltag2_clone_only,normalization.method = "LogNormalize", scale.factor =10000)

```



# Fetal 2D and 3D plus the multiciliated extraction 
```{r}

fetal.3D.one <- CreateSeuratObject(counts = fetal.3D.one.data, project = "fetal.3D.one", min.cells = 3, min.features = 200)
fetal.3D.two <- CreateSeuratObject(counts = fetal.3D.two.data, project = "fetal.3D.two", min.cells = 3, min.features = 200)
fetal.3D.one <- RenameCells(object = fetal.3D.one, add.cell.id = "fetal.3D.one")
fetal.3D.two <- RenameCells(object = fetal.3D.two, add.cell.id = "fetal.3D.two")

fetal.ali.one <- CreateSeuratObject(counts = fetal.ali.one.data, project = "fetal.ali.one", min.cells = 3, min.features = 200)
fetal.ali.two <- CreateSeuratObject(counts = fetal.ali.two.data, project = "fetal.ali.two", min.cells = 3, min.features = 200)
fetal.ali.one <- RenameCells(object = fetal.ali.one, add.cell.id = "fetal.ali.one")
fetal.ali.two <- RenameCells(object = fetal.ali.two, add.cell.id = "fetal.ali.two")



final_seurat_object <- merge(x = fetal.3D.one, y = c(fetal.3D.two,fetal.ali.one,fetal.ali.two))


sample_species <- "human"
qc_max_mito <- ifelse(sample_species == "human", 10, 5)
final_seurat_object[["percent.mt"]] <- ifelse(sample_species == "human",
                                                  PercentageFeatureSet(final_seurat_object, pattern = "^MT-"),
                                                  PercentageFeatureSet(final_seurat_object, pattern = "^mt-"))


VlnPlot(final_seurat_object, features <- c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

qc_min_genes= 200
qc_min_counts = 200
qc_max_genes = 12000
qc_max_counts = 120000


final_seurat_object <- subset(final_seurat_object, subset = nFeature_RNA >= qc_min_genes &
                                nFeature_RNA <= qc_max_genes & percent.mt <= qc_max_mito &
                                nCount_RNA >= qc_min_counts & nCount_RNA <= qc_max_counts)

## batch correction
final_seurat_object <- SplitObject(final_seurat_object, split.by = "orig.ident")

set.seed(888)
for (i in 1:length(final_seurat_object)) {
    final_seurat_object[[i]] <- SCTransform(final_seurat_object[[i]],
                                            verbose = FALSE, vars.to.regress = "percent.mt")
}

final_seurat_object_features <- SelectIntegrationFeatures(object.list = final_seurat_object,nfeatures = 3000)
final_seurat_object <- PrepSCTIntegration(object.list = final_seurat_object,
                                          anchor.features = final_seurat_object_features)

final_seurat_object_anchors <- FindIntegrationAnchors(object.list = final_seurat_object,
                                           normalization.method = "SCT",
                                           anchor.features = final_seurat_object_features)



final_seurat_object_integrated <- IntegrateData(anchorset = final_seurat_object_anchors,
                                     normalization.method = "SCT")


set.seed(888)
final_seurat_object_integrated <- RunPCA(final_seurat_object_integrated, verbose = FALSE)
final_seurat_object_integrated <- RunUMAP(final_seurat_object_integrated, dims = 1:30,return.model=TRUE)
final_seurat_object_integrated <- FindNeighbors(final_seurat_object_integrated, dims = 1:30)
final_seurat_object_integrated <- FindClusters(final_seurat_object_integrated, resolution = 0.3, algorithm = 4)


set.seed(888)
DefaultAssay(final_seurat_object_integrated) <- "RNA"
final_seurat_object_integrated <- NormalizeData(final_seurat_object_integrated,
                                            normalization.method = "LogNormalize", scale.factor =10000)


# cluster 6 extraction

set.seed(888)
final_seurat_object_integrated6 <- subset(x = final_seurat_object_integrated, idents = c(6))
final_seurat_object_integrated6 <- SCTransform(final_seurat_object_integrated6,
                                        vars.to.regress = "percent.mt", verbose = FALSE)

final_seurat_object_integrated6 <- RunPCA(final_seurat_object_integrated6, verbose = FALSE)

final_seurat_object_integrated6 <- RunUMAP(final_seurat_object_integrated6, dims = 1:30)
final_seurat_object_integrated6 <- FindNeighbors(final_seurat_object_integrated6, dims = 1:30)
final_seurat_object_integrated6 <- FindClusters(final_seurat_object_integrated6, resolution = 0.5, algorithm = 4)

DefaultAssay(final_seurat_object_integrated6) <- "RNA"
final_seurat_object_integrated6 <- NormalizeData(final_seurat_object_integrated6,
                                            normalization.method = "LogNormalize", scale.factor =10000)


# cell scoring plots

fetal.fasl.0 <- list(c("SCGB3A2",
                       "SFTPB",
                       "CYB5A",
                       "VIM",
                       "C16orf89",
                       "CXCL17",
                       "XBP1",
                       "MGST1",
                       "SSR4",
                       "CFH",
                       "LMO3",
                       "C3",
                       "STEAP4",
                       "CLU",
                       "FMO2",
                       "CP",
                       "CFTR",
                       "KLK11",
                       "CYTL1",
                       "HNMT",
                       "CFHR1",
                       "SLC4A4",
                       "HOPX",
                       "RNASE1",
                       "SERPINF1",
                       "TMSB4X",
                       "NDRG1",
                       "KLHL24",
                       "GDF15",
                       "KIAA1324",
                       "CST3",
                       "TMEM45A",
                       "SERPINA1",
                       "MET",
                       "PTP4A2",
                       "SERPINI1",
                       "SPINK5",
                       "SOD3",
                       "PTP4A1",
                       "MUC5B",
                       "IFITM3",
                       "SFTA2",
                       "IFITM2",
                       "TFF3",
                       "RAMP2",
                       "ASAH1",
                       "DSTN",
                       "RNU12",
                       "SLC39A6",
                       "FAM46C"))
fetal.btp.1 <- list(c("SFTPC",
                      "NPC2",
                      "TESC",
                      "CA2",
                      "ETV5",
                      "ASPSCR1",
                      "WIF1",
                      "CD36",
                      "CPM",
                      "SFTA3",
                      "IGFBP7",
                      "TPD52L1",
                      "MFSD2A",
                      "ATP11A",
                      "TPM1",
                      "FTL",
                      "SLC34A2",
                      "BAMBI",
                      "ID2",
                      "COL9A3",
                      "TKT",
                      "HNRNPA1",
                      "ADAMTS1",
                      "PIK3C2G",
                      "NHSL1",
                      "TMSB10",
                      "NREP",
                      "MARCKSL1",
                      "NOTUM",
                      "CLDN6",
                      "MEST",
                      "NPM1",
                      "GSTM3",
                      "BTG1",
                      "RPL5",
                      "RPS23",
                      "NSG1",
                      "RPL6",
                      "ENO1",
                      "CLDN2",
                      "RPS3A",
                      "EDNRB",
                      "RPL3",
                      "RPL11",
                      "EIF3E",
                      "HAS3",
                      "EEF1A1",
                      "MYL9",
                      "LDHB",
                      "SFTPA1"))
fetal.bta.2 <- list(c("AGER",
                    "CLIC3",
                    "MYL9",
                    "TNNC1",
                    "CAV1",
                    "SMARCA5",
                    "S100A10",
                    "CLDN6",
                    "ICAM4",
                    "CLDN18",
                    "SFTA3",
                    "FOLR1",
                    "AQP4",
                    "SPARC",
                    "EDN3",
                    "NDNF",
                    "PCP4",
                    "NGFRAP1",
                    "PDPN",
                    "MXRA8",
                    "TMSB10",
                    "ANKRD29",
                    "FTL",
                    "RPS20",
                    "RPS3",
                    "RPL5",
                    "MBIP",
                    "CD47",
                    "TUBB",
                    "ANXA3",
                    "ZFAS1",
                    "PTMA",
                    "NKX2-1",
                    "RPL31",
                    "TSPAN13",
                    "RPL11",
                    "GDF15",
                    "RPL35A",
                    "RPL23A",
                    "RPS27",
                    "RPL41",
                    "RPS8",
                    "RPL10A",
                    "NREP",
                    "RPS23",
                    "NPM1",
                    "RPL24",
                    "RPS25",
                    "HOPX"))
fetal.multiciliated.3 <- list(c("TMEM190",
                                "CAPS",
                                "C9orf24",
                                "C20orf85",
                                "C1orf194",
                                "FAM183A",
                                "RP11-356K23.1",
                                "OMG",
                                "RSPH1",
                                "PIFO",
                                "AGR3",
                                "ODF3B",
                                "SNTN",
                                "FABP6",
                                "TSPAN1",
                                "CCDC78",
                                "DYNLRB2",
                                "DNAAF1",
                                "TPPP3",
                                "C5orf49",
                                "MS4A8",
                                "C9orf116",
                                "CAPSL",
                                "SMIM22",
                                "C9orf117",
                                "CETN2",
                                "AC013264.2",
                                "MORN5",
                                "TSPAN19",
                                "TUBB4B",
                                "CRIP1",
                                "CDHR3",
                                "MRPS31",
                                "MORN2",
                                "C1orf192",
                                "FAM92B",
                                "PSENEN",
                                "FAM216B",
                                "TMC5",
                                "ZMYND10",
                                "CCDC17",
                                "LRRIQ1",
                                "C22orf15",
                                "C21orf58",
                                "C11orf88",
                                "EFHC1",
                                "SLC44A4",
                                "LRRC46",
                                "CES1",
                                "FOXJ1"))
fetal.basal.4 <- list(c("KRT15",
                        "KRT5",
                        "S100A2",
                        "MIR205HG",
                        "AQP3",
                        "KRT19",
                        "TACSTD2",
                        "SFN",
                        "JUNB",
                        "KLF5",
                        "PERP",
                        "F3",
                        "CD9",
                        "HCAR3",
                        "IGFBP2",
                        "JUN",
                        "CSTA",
                        "NPPC",
                        "HCAR2",
                        "FOS",
                        "IL33",
                        "DLK2",
                        "ADH7",
                        "CAPNS2",
                        "PKP1",
                        "TP63",
                        "PNCK",
                        "RPLP1",
                        "BHLHE40",
                        "HSPB1",
                        "ZFP36",
                        "GJB2",
                        "KRT13",
                        "LSP1",
                        "EGR1",
                        "IGFBP3",
                        "SERPINB5",
                        "IER3",
                        "ARL4D",
                        "MYC",
                        "MPZL2",
                        "COL7A1",
                        "PLP2",
                        "MEG3",
                        "ZFP36L1",
                        "CALML3",
                        "FOSB",
                        "LMNA",
                        "PVRL1",
                        "ACKR3"))
fetal.secretory1.5 <- list(c("SCGB1A1",
                             "KRT4",
                             "WFDC2",
                             "MSLN",
                             "IGF2",
                             "SERPINB3",
                             "SLPI",
                             "ANXA1",
                             "MUC4",
                             "S100A9",
                             "KRT13",
                             "S100P",
                             "KRT19",
                             "TACSTD2",
                             "UPK1B",
                             "IGFBP3",
                             "S100A6",
                             "AQP3",
                             "CLDN4",
                             "APOBEC3A",
                             "CYP2F1",
                             "RHOV",
                             "TNNT3",
                             "CEACAM6",
                             "C19orf33",
                             "PLAC8",
                             "KLF5",
                             "ELF3",
                             "FAM3D",
                             "LSP1",
                             "VMO1",
                             "F3",
                             "LCN2",
                             "LYPD2",
                             "CSTA",
                             "ASS1",
                             "SLC16A9",
                             "TNFSF10",
                             "SERPINB2",
                             "GPR110",
                             "MUC20",
                             "EPHA2",
                             "TCN1",
                             "A4GALT",
                             "GABRP",
                             "LYN",
                             "PDE4C",
                             "FUT3",
                             "TFF3",
                             "FABP5"))
fetal.secretory2.6 <- list(c("LTF",
                              "TM4SF1",
                              "SLPI",
                              "KRT7",
                              "AZGP1",
                              "NDRG2",
                              "TTYH1",
                              "MIA",
                              "SCGB3A1",
                              "WFDC2",
                              "APIP",
                              "CALML5",
                              "PHLDA1",
                              "SLC12A2",
                              "FAM3D",
                              "TCN1",
                              "EHF",
                              "CCL28",
                              "KRT14",
                              "MYC",
                              "SHISA2",
                              "MGP",
                              "GMDS",
                              "FCGBP",
                              "FABP7",
                              "KCNN4",
                              "PTN",
                              "FOXC1",
                              "GLYATL2",
                              "ZG16B",
                              "NFIB",
                              "APOE",
                              "SOX9",
                              "AQP5",
                              "RPL36",
                              "ALDH1A3",
                              "SERPINE2",
                              "CTNNB1",
                              "DUSP6",
                              "TIMP1",
                              "COL9A3",
                              "INSR",
                              "KRT19",
                              "EFCAB4A",
                              "TSPAN8",
                              "ZFP36L1",
                              "GABRP",
                              "EPHB3",
                              "NFKBIZ",
                              "PABPC1"))
fetal.multiciliatedprecursor.7 <- list(c("CDC20B",
                                       "LRRC26",
                                       "HES6",
                                       "ZMYND10",
                                       "CDK1",
                                       "CCDC74A",
                                       "CCDC67",
                                       "HYLS1",
                                       "NEK2",
                                       "CDC20",
                                       "C1orf192",
                                       "PLK4",
                                       "SPAG6",
                                       "RIBC2",
                                       "CCDC19",
                                       "ANLN",
                                       "POC1A",
                                       "HIST1H2BJ",
                                       "FOXN4",
                                       "E2F7",
                                       "MCIDAS",
                                       "SYT5",
                                       "STIL",
                                       "C11orf88",
                                       "CCNO",
                                       "C5orf49",
                                       "MELK",
                                       "ROPN1L",
                                       "RP11-263K19.4",
                                       "C11orf16",
                                       "SGOL2",
                                       "MUC12",
                                       "FOXJ1",
                                       "FAM183A",
                                       "C7orf57",
                                       "TEKT2",
                                       "DNAH12",
                                       "RSPH9",
                                       "STMND1",
                                       "LRRC46",
                                       "DNAAF3",
                                       "PITPNM1",
                                       "KDELC2",
                                       "RSPH1",
                                       "ARMC3",
                                       "CEP152",
                                       "MYB",
                                       "PIFO",
                                       "SPEF1",
                                       "CAPSL"))
fetal.neuroendocrine.8 <- list(c("GHRL",
                                 "GRP",
                                 "SEC11C",
                                 "PCSK1N",
                                 "CPE",
                                 "NNAT",
                                 "CHGA",
                                 "ASCL1",
                                 "RPRM",
                                 "IGFBP5",
                                 "HES6",
                                 "BEX1",
                                 "ACSL1",
                                 "SCGN",
                                 "HEPACAM2",
                                 "CALCA",
                                 "MEG3",
                                 "TMEM176B",
                                 "TMEM176A",
                                 "BIK",
                                 "SCG5",
                                 "APOA1BP",
                                 "UCHL1",
                                 "MIAT",
                                 "C12orf75",
                                 "KLK12",
                                 "CHGB",
                                 "SLC35D3",
                                 "NPDC1",
                                 "DDC",
                                 "TAGLN3",
                                 "GFRA3",
                                 "SCG3",
                                 "NPW",
                                 "SCG2",
                                 "TUBB2B",
                                 "C3orf14",
                                 "ATP6V0B",
                                 "MAOB",
                                 "VAMP2",
                                 "C4orf48",
                                 "APLP1",
                                 "LINC00261",
                                 "BEX2",
                                 "BCAM",
                                 "LY6H",
                                 "CA8",
                                 "FAM105A",
                                 "TUBA1A",
                                 "HOXB5"))



set.seed(888)
final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.fasl.0, 
                                     name = "fetal.fasl.0", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.btp.1, 
                                     name = "fetal.btp.1", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.bta.2, 
                                     name = "fetal.bta.2", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.multiciliated.3, 
                                     name = "fetal.multiciliated.3", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.basal.4, 
                                     name = "fetal.basal.4", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.secretory1.5, 
                                     name = "fetal.secretory1.5", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.secretory2.6, 
                                     name = "fetal.secretory2.6", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.multiciliatedprecursor.7, 
                                     name = "fetal.multiciliatedprecursor.7", search = TRUE, assay = "RNA")

final_seurat_object_integrated <-AddModuleScore(final_seurat_object_integrated, features = fetal.neuroendocrine.8, 
                                     name = "fetal.neuroendocrine.8", search = TRUE, assay = "RNA")




FeaturePlot(final_seurat_object_integrated, features = "fetal.fasl.01",raster = FALSE,
            slot = "data",pt.size = 0.5) & RotatedAxis() & scale_colour_gradientn(colors =                                                                                  c("lightgrey","#ffffd9","#edf8b1","#c7e9b4","#7fcdbb","#41b6c4","#1d91c0","#225ea8","#253494","#081d58")) &   xlab("UMAP1") & ylab("UMAP2") & labs(title = "fetal.fasl.0")
```

